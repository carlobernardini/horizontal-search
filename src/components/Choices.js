import React, { Component } from 'react';
import PropTypes from 'prop-types';
import {
    includes, isEqual, uniqueId, without
} from 'lodash-es';
import { DISPLAY_NAME_FILTER_CHOICES, NAME_PREFIX_CHECKBOX, NAME_PREFIX_RADIO } from '../constants';
import CheckBox from './CheckBox';
import Caption from './Caption';
import FilterInfo from './FilterInfo';
import RadioButton from './RadioButton';
import '../scss/Filter.scss';

class Choices extends Component {
    renderChoices() {
        const {
            fieldName, multiple, onUpdate, options, value
        } = this.props;

        if (multiple) {
            return options.map((option, i) => (
                <CheckBox
                    key={ i }
                    id={ uniqueId(NAME_PREFIX_CHECKBOX) }
                    disabled={ option.disabled }
                    facetCount={ option.facetCount }
                    value={ option.value }
                    selected={ (value || []).indexOf(option.value) > -1 }
                    label={ option.label || option.value }
                    onChange={ (selected) => {
                        let newValue = (value || []).slice(0);
                        if (selected && !includes(newValue, option.value)) {
                            newValue.push(option.value);
                        }
                        if (!selected && includes(newValue, option.value)) {
                            newValue = without(newValue, option.value);
                        }
                        if (isEqual(newValue, value)) {
                            return false;
                        }
                        onUpdate(newValue.length ? newValue : null);
                        return true;
                    }}
                />
            ));            
        }

        return options.map((option, i) => (
            <RadioButton
                key={ i }
                id={ uniqueId(NAME_PREFIX_RADIO) }
                name={ fieldName }
                disabled={ option.disabled }
                facetCount={ option.facetCount }
                value={ option.value }
                selected={ value === option.value }
                label={ option.label || option.value }
                onChange={ (selectedValue) => (onUpdate(selectedValue)) }
            />
        ));
    }

    renderNoPreference() {
        const { noPreferenceLabel, multiple } = this.props;

        if (multiple || !noPreferenceLabel) {
            return null;
        }

        const { fieldName, onUpdate, value } = this.props;

        return (
            <RadioButton
                id={ uniqueId(NAME_PREFIX_RADIO) }
                name={ fieldName }
                value="no-preference"
                selected={ !value }
                label={ noPreferenceLabel }
                onChange={ () => (onUpdate(null)) }
            />
        );
    }

    render() {
        const { information, label } = this.props;
        return (
            <div
                className="orizzonte__filter"
            >
                <FilterInfo
                    information={ information }
                />
                <Caption>
                    { label }
                </Caption>
                { this.renderNoPreference() }
                { this.renderChoices() }
            </div>
        );
    }
}

Choices.displayName = DISPLAY_NAME_FILTER_CHOICES;

Choices.propTypes = {
    /** Field name for this filter, to be used in composed query */
    fieldName: PropTypes.string.isRequired,
    information: PropTypes.string,
    /** Label for this filter section */
    label: PropTypes.string,
    /** Whether to show checkboxes (true) or radios (false) */
    multiple: PropTypes.bool,
    /** Label to show if you want to include a 'no preference' option
        Only available for radio groups */
    noPreferenceLabel: PropTypes.string,
    /** Internal callback for when filter value has changed */
    onUpdate: PropTypes.func,
    /** List of selectable options (value is required) */
    options: PropTypes.arrayOf(
        PropTypes.shape({
            disabled: PropTypes.bool,
            facetCount: PropTypes.oneOfType([
                PropTypes.string,
                PropTypes.number
            ]),
            label: PropTypes.string,
            value: PropTypes.oneOfType([
                PropTypes.number,
                PropTypes.string
            ]).isRequired
        })
    ).isRequired,
    value: PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.number,
        PropTypes.arrayOf(
            PropTypes.oneOfType([
                PropTypes.string,
                PropTypes.number
            ])
        )
    ])
};

Choices.defaultProps = {
    information: null,
    label: null,
    multiple: false,
    noPreferenceLabel: null,
    onUpdate: () => {},
    value: null
};

export default Choices;
