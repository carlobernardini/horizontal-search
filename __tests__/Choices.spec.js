import React from 'react';
import Choices from '../src/components/Choices';

describe('<Choices />', () => {
    it('should render a series of checkboxes with some disabled ones', () => {
        const wrapper = shallow(
            <Choices
                fieldName="myAPIField"
                label="Test multiple choices"
                options={ [{
                    label: 'Test value 1',
                    value: 1,
                    disabled: true
                }, {
                    label: 'Test value 2',
                    value: 2
                }, {
                    label: 'Test value 3',
                    value: 3,
                    disabled: true
                }]}
                onUpdate={() => {}}
                multiple
            />
        );

        expect(wrapper).toMatchSnapshot();
    });

    it('should render a series of checkboxes with selected values', () => {
        const wrapper = shallow(
            <Choices
                fieldName="myAPIField"
                label="Test multiple choices"
                options={ [{
                    label: 'Test value 1',
                    value: 1
                }, {
                    label: 'Test value 2',
                    value: 2
                }, {
                    label: 'Test value 3',
                    value: 3
                }]}
                onUpdate={() => {}}
                value={[2, 3]}
                multiple
            />
        );

        expect(wrapper).toMatchSnapshot();
    });

    it('should render a series of radios with some disabled ones', () => {
        const wrapper = shallow(
            <Choices
                fieldName="myAPIField"
                label="Test multiple choices"
                options={ [{
                    label: 'Test value 1',
                    value: 1,
                    disabled: true
                }, {
                    label: 'Test value 2',
                    value: 2
                }, {
                    label: 'Test value 3',
                    value: 3,
                    disabled: true
                }]}
                onUpdate={() => {}}
            />
        );

        expect(wrapper).toMatchSnapshot();
    });

    it('should render a series of radios with selected value', () => {
        const wrapper = shallow(
            <Choices
                fieldName="myAPIField"
                label="Test multiple choices"
                options={ [{
                    label: 'Test value 1',
                    value: 1,
                    disabled: true
                }, {
                    label: 'Test value 2',
                    value: 2
                }, {
                    label: 'Test value 3',
                    value: 3,
                    disabled: true
                }]}
                onUpdate={() => {}}
                value={ 3 }
            />
        );

        expect(wrapper).toMatchSnapshot();
    });
});
