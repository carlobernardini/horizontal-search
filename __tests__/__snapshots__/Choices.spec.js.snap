// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Choices /> should render a series of checkboxes with selected values 1`] = `
<div
  className="orizzonte__filter"
>
  <OrizzonteFilterInfo
    information={null}
  />
  <div
    className="orizzonte__filter-caption"
  >
    Test multiple choices
  </div>
  <OrizzonteCheckBox
    disabled={false}
    id="checkbox-4"
    key="0"
    label="Test value 1"
    onChange={[Function]}
    selected={false}
    value={1}
    viewBox={null}
  />
  <OrizzonteCheckBox
    disabled={false}
    id="checkbox-5"
    key="1"
    label="Test value 2"
    onChange={[Function]}
    selected={true}
    value={2}
    viewBox={null}
  />
  <OrizzonteCheckBox
    disabled={false}
    id="checkbox-6"
    key="2"
    label="Test value 3"
    onChange={[Function]}
    selected={true}
    value={3}
    viewBox={null}
  />
</div>
`;

exports[`<Choices /> should render a series of checkboxes with some disabled ones 1`] = `
<div
  className="orizzonte__filter"
>
  <OrizzonteFilterInfo
    information={null}
  />
  <div
    className="orizzonte__filter-caption"
  >
    Test multiple choices
  </div>
  <OrizzonteCheckBox
    disabled={true}
    id="checkbox-1"
    key="0"
    label="Test value 1"
    onChange={[Function]}
    selected={false}
    value={1}
    viewBox={null}
  />
  <OrizzonteCheckBox
    disabled={false}
    id="checkbox-2"
    key="1"
    label="Test value 2"
    onChange={[Function]}
    selected={false}
    value={2}
    viewBox={null}
  />
  <OrizzonteCheckBox
    disabled={true}
    id="checkbox-3"
    key="2"
    label="Test value 3"
    onChange={[Function]}
    selected={false}
    value={3}
    viewBox={null}
  />
</div>
`;

exports[`<Choices /> should render a series of radios with selected value 1`] = `
<div
  className="orizzonte__filter"
>
  <OrizzonteFilterInfo
    information={null}
  />
  <div
    className="orizzonte__filter-caption"
  >
    Test multiple choices
  </div>
  <OrizzonteRadioButton
    disabled={true}
    id="radio-10"
    key="0"
    label="Test value 1"
    name="myAPIField"
    onChange={[Function]}
    selected={false}
    value={1}
  />
  <OrizzonteRadioButton
    disabled={false}
    id="radio-11"
    key="1"
    label="Test value 2"
    name="myAPIField"
    onChange={[Function]}
    selected={false}
    value={2}
  />
  <OrizzonteRadioButton
    disabled={true}
    id="radio-12"
    key="2"
    label="Test value 3"
    name="myAPIField"
    onChange={[Function]}
    selected={true}
    value={3}
  />
</div>
`;

exports[`<Choices /> should render a series of radios with some disabled ones 1`] = `
<div
  className="orizzonte__filter"
>
  <OrizzonteFilterInfo
    information={null}
  />
  <div
    className="orizzonte__filter-caption"
  >
    Test multiple choices
  </div>
  <OrizzonteRadioButton
    disabled={true}
    id="radio-7"
    key="0"
    label="Test value 1"
    name="myAPIField"
    onChange={[Function]}
    selected={false}
    value={1}
  />
  <OrizzonteRadioButton
    disabled={false}
    id="radio-8"
    key="1"
    label="Test value 2"
    name="myAPIField"
    onChange={[Function]}
    selected={false}
    value={2}
  />
  <OrizzonteRadioButton
    disabled={true}
    id="radio-9"
    key="2"
    label="Test value 3"
    name="myAPIField"
    onChange={[Function]}
    selected={false}
    value={3}
  />
</div>
`;
